platform/api/aws/images.go

```diff
@@ -0,0 +1,229 @@
+	Type EC2ImageType
+	// Image region
+	Region string
+}
+
+type Snapshot struct {
+	SnapshotID string
+}
+
+// CreateSnapshot creates an AWS Snapshot
+func (a *API) CreateSnapshot(description, sourceURL string) (*Snapshot, error) {
+	s3url, err := url.Parse(sourceURL)
+	if err != nil {
+		return nil, err
+	}
+	if s3url.Scheme != "s3" {
+		return nil, fmt.Errorf("source must have a 's3://' scheme, not: '%v://'", s3url.Scheme)
+	}
+	s3key := strings.TrimPrefix(s3url.Path, "/")
+
+	importRes, err := a.ec2.ImportSnapshot(&ec2.ImportSnapshotInput{
+		RoleName:    aws.String(vmImportRole),
+		Description: aws.String(description),
+		DiskContainer: &ec2.SnapshotDiskContainer{
+			// TODO(euank): allow s3 source / local file -> s3 source
+			UserBucket: &ec2.UserBucket{
+				S3Bucket: aws.String(s3url.Host),
+				S3Key:    aws.String(s3key),
+			},
+			Format: aws.String(ec2.DiskImageFormatVmdk),
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create import snapshot task: %v", err)
+	}
+
+	plog.Infof("created snapshot import task %v", *importRes.ImportTaskId)
+
+	snapshotDone := func(snapshotTaskID string) (bool, string, error) {
+		taskRes, err := a.ec2.DescribeImportSnapshotTasks(&ec2.DescribeImportSnapshotTasksInput{
+			ImportTaskIds: []*string{aws.String(snapshotTaskID)},
+		})
+		if err != nil {
+			return false, "", err
+		}
+
+		details := taskRes.ImportSnapshotTasks[0].SnapshotTaskDetail
+
+		// I dream of AWS specifying this as an enum shape, not string
+		switch *details.Status {
+		case "completed":
+			return true, *details.SnapshotId, nil
+		case "pending", "active":
+			plog.Debugf("waiting for import task: %v (%v): %v", *details.Status, *details.Progress, *details.StatusMessage)
+			return false, "", nil
+		case "cancelled", "cancelling":
+			return false, "", fmt.Errorf("import task cancelled")
+		case "deleted", "deleting":
+			errMsg := "unknown error occured importing snapshot"
+			if details.StatusMessage != nil {
+				errMsg = *details.StatusMessage
+			}
+			return false, "", fmt.Errorf("could not import snapshot: %v", errMsg)
+		default:
+			return false, "", fmt.Errorf("unexpected status: %v", *details.Status)
+		}
+	}
+
+	// TODO(euank): write a waiter for import snapshot
+	for {
+		done, snapshotID, err := snapshotDone(*importRes.ImportTaskId)
+		if err != nil {
+			return nil, err
+		}
+		if done {
+			return &Snapshot{
+				SnapshotID: snapshotID,
+			}, nil
+		}
+		time.Sleep(20 * time.Second)
+	}
+}
+
+// CreateHVMImage creates an HVM AMI
+// The image will be created in the region configured for this API.
+func (a *API) CreateHVMImage(name, description, snapshot *Snapshot) (Image, error) {
+	a.ec2.RegisterImage(&ec2.RegisterImageInput{})
+	return Image{}, nil
+}
+
+// CreateHVMImage creates a PV AMI
+// The image will be created in the region configured for this API.
+func (a *API) CreatePVImage(name, description, snapshot *Snapshot) (Image, error) {
+	a.ec2.RegisterImage(&ec2.RegisterImageInput{})
```

Comment:
Preference for whether I nix em or finish em up in this PR? Basically, is this approaching "too big"?